File: lc/T1-50/t_40.java
package leetcoder;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;


class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> list = new ArrayList<>();
        if(candidates == null || candidates.length ==0 ) return list;
        Arrays.sort(candidates);
        dfs(list,new ArrayList<Integer>(),candidates,target,0,0);
        return list;
    }
    void dfs(List<List<Integer>> list ,List<Integer> tmp,int []a,int t,int sum,int now){
        if(t==sum){
            list.add(new ArrayList<>(tmp));
            return;
        }else if( now > a.length || ((now < a.length)&& t-sum < a[now])){
            return;
        }
        for(int i=now;i<a.length;i++){
            tmp.add(a[i]);
            dfs(list,tmp,a,t,sum+a[i],i+1);
            tmp.remove(tmp.size()-1);
            while(i+1<a.length && a[i+1]==a[i])i++;
        }
    }
}

File: lc/T1-50/T40.cc


#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
    {
        set<vector<int>> res;
        vector<int> tmp;
        dfs(0, 0, target, res, candidates, tmp);  

        return vector<vector<int>> (res.begin (), res.end ());


        
    }
    void dfs(int now, int score, int target,set<vector<int>> &res, vector<int> &candidates, vector<int> v)
    {
        if (score == target)
        {
            vector<int> x = v;
            sort(x.begin (), x.end());
            res.insert(x);
            return;
        }
        if (now == candidates.size())
            return;
        dfs(now + 1, score, target, res, candidates, v);
        v.push_back(candidates[now]);
        dfs(now + 1, score + candidates[now], target, res, candidates,v);
    }
};

int main()
{ 

    vector<int> v{10, 1, 2, 7, 6, 1, 5};
    Solution s;
    auto it = s.combinationSum2(v, 8);
    for (auto i : it)
    {
        for (auto j : i)
        {
            cout << j << " ";
        }
        cout << endl;
    }
    return 0;
}



// [[7,1],[2,1,5],[2,6],[1,6,1],[1,7],[1,2,5]]
// [[1,1,6],[1,2,5],[1,7],[2,6]]